This notebook shows the data preparation process to the final dataset for use.

The first dataset we decided to use was the Purchasing behavior in January(found [link](https://www.kaggle.com/mkechinov/ecommerce-events-history-in-cosmetics-shop)). I thought this would be a good estimate of purchasing behavior in cosmetics product in general.

**The attributes I decided to focus on are:**

1. event_type

2. event_time

3. brand

4. price

5. category_code




**Prepare Raw Data**

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from datetime import date
import seaborn as sns


cos = pd.read_csv('2020-Jan.csv')
cos.head()

event_time event_type  product_id          category_id  \
0  2020-01-01 00:00:00 UTC       view     5809910  1602943681873052386   
1  2020-01-01 00:00:09 UTC       view     5812943  1487580012121948301   
2  2020-01-01 00:00:19 UTC       view     5798924  1783999068867920626   
3  2020-01-01 00:00:24 UTC       view     5793052  1487580005754995573   
4  2020-01-01 00:00:25 UTC       view     5899926  2115334439910245200   

  category_code     brand  price    user_id  \
0           NaN   grattol   5.24  595414620   
1           NaN  kinetics   3.97  595414640   
2           NaN    zinger   3.97  595412617   
3           NaN       NaN   4.92  420652863   
4           NaN       NaN   3.92  484071203   

                           user_session  
0  4adb70bb-edbd-4981-b60f-a05bfd32683a  
1  c8c5205d-be43-4f1d-aa56-4828b8151c8a  
2  46a5010f-bd69-4fbe-a00d-bb17aa7b46f3  
3  546f6af3-a517-4752-a98b-80c4c5860711  
4  cff70ddf-529e-4b0c-a4fc-f43a749c0acb

Drop Empty value in event_time, event_type, brand, price. This is because these four factors are the categories where the research wants to focus in. By dropping empy value, only purchases with defined label will be analyzed.

The process of dropping empty values and display the final dataset

In [3]:
cleaning=cos.copy()

#cleaning= cleaning.dropna(subset=['category_code'])

cleaning= cleaning.dropna(subset=['price'])

cos= cos.dropna(subset=['brand'])

cleaning= cleaning.dropna(subset=['event_type'])

cleaning= cleaning.dropna(subset=['event_time'])

cleaning.head()


droppingall=cos.copy()
droppingall = droppingall.dropna(subset=['category_code'])

cleaning.head()

event_time event_type  product_id          category_id  \
0  2020-01-01 00:00:00 UTC       view     5809910  1602943681873052386   
1  2020-01-01 00:00:09 UTC       view     5812943  1487580012121948301   
2  2020-01-01 00:00:19 UTC       view     5798924  1783999068867920626   
3  2020-01-01 00:00:24 UTC       view     5793052  1487580005754995573   
4  2020-01-01 00:00:25 UTC       view     5899926  2115334439910245200   

  category_code     brand  price    user_id  \
0           NaN   grattol   5.24  595414620   
1           NaN  kinetics   3.97  595414640   
2           NaN    zinger   3.97  595412617   
3           NaN       NaN   4.92  420652863   
4           NaN       NaN   3.92  484071203   

                           user_session  
0  4adb70bb-edbd-4981-b60f-a05bfd32683a  
1  c8c5205d-be43-4f1d-aa56-4828b8151c8a  
2  46a5010f-bd69-4fbe-a00d-bb17aa7b46f3  
3  546f6af3-a517-4752-a98b-80c4c5860711  
4  cff70ddf-529e-4b0c-a4fc-f43a749c0acb

In [4]:
print("The length of dataset after cleaning:" +str(len(cleaning)))
print(" ")
print("If dropping all NaN values, The length of non-empty dataset:" +str(len(droppingall)))


The length of dataset after cleaning:4264752
 
If dropping all NaN values, The length of non-empty dataset:57782


I didn't drop NaN in category because if I drop the NaN in category_code. The length of non-empty dataset would be 57782, that is significantly fewer than the raw dataset with 4262752 raw data and the 2489122 data(without dropping the category code). In order to preserve most of the data collected, and for better accuracy of the result, I would not drop the NaN value in price and category_code at this stage.

**Preparing dummies variable**

In [5]:
cos = cleaning
cos['cart']=pd.get_dummies(cos['event_type'])['cart']
cos['purchase']=pd.get_dummies(cos['event_type'])['purchase']
cos['remove_from_cart']=pd.get_dummies(cos['event_type'])['remove_from_cart']
cos['view']=pd.get_dummies(cos['event_type'])['view']
pd.get_dummies(cos['event_type']).sample(10)


cart  purchase  remove_from_cart  view
461819      0         0                 0     1
1677709     1         0                 0     0
2562773     0         0                 1     0
3367001     0         0                 1     0
1234681     0         0                 1     0
1516743     0         0                 1     0
1104654     0         1                 0     0
1068494     0         0                 0     1
351994      0         0                 0     1
3037065     0         0                 0     1

In [6]:
cos.head()


event_time event_type  product_id          category_id  \
0  2020-01-01 00:00:00 UTC       view     5809910  1602943681873052386   
1  2020-01-01 00:00:09 UTC       view     5812943  1487580012121948301   
2  2020-01-01 00:00:19 UTC       view     5798924  1783999068867920626   
3  2020-01-01 00:00:24 UTC       view     5793052  1487580005754995573   
4  2020-01-01 00:00:25 UTC       view     5899926  2115334439910245200   

  category_code     brand  price    user_id  \
0           NaN   grattol   5.24  595414620   
1           NaN  kinetics   3.97  595414640   
2           NaN    zinger   3.97  595412617   
3           NaN       NaN   4.92  420652863   
4           NaN       NaN   3.92  484071203   

                           user_session  cart  purchase  remove_from_cart  \
0  4adb70bb-edbd-4981-b60f-a05bfd32683a     0         0                 0   
1  c8c5205d-be43-4f1d-aa56-4828b8151c8a     0         0                 0   
2  46a5010f-bd69-4fbe-a00d-bb17aa7b46f3     0         0                 0   
3  546f6af3-a517-4752-a98b-80c4c5860711     0         0                 0   
4  cff70ddf-529e-4b0c-a4fc-f43a749c0acb     0         0                 0   

   view  
0     1  
1     1  
2     1  
3     1  
4     1

**changing datatype of 'price'**

In [7]:
cos['price']

0          5.24
1          3.97
2          3.97
3          4.92
4          3.92
           ... 
4264747    1.59
4264748    1.59
4264749    2.05
4264750    2.22
4264751    2.22
Name: price, Length: 4264752, dtype: float64

In [8]:
cos['price'] = pd.to_numeric(cos['price'])

type(cos['price'][0])

numpy.float64

**Formalities time conversion**

In [9]:
%time
cos['event_time'] = pd.to_datetime(cos['event_time'],infer_datetime_format=True)
cos.head()

CPU times: user 2 µs, sys: 0 ns, total: 2 µs
Wall time: 4.05 µs


event_time event_type  product_id          category_id  \
0 2020-01-01 00:00:00+00:00       view     5809910  1602943681873052386   
1 2020-01-01 00:00:09+00:00       view     5812943  1487580012121948301   
2 2020-01-01 00:00:19+00:00       view     5798924  1783999068867920626   
3 2020-01-01 00:00:24+00:00       view     5793052  1487580005754995573   
4 2020-01-01 00:00:25+00:00       view     5899926  2115334439910245200   

  category_code     brand  price    user_id  \
0           NaN   grattol   5.24  595414620   
1           NaN  kinetics   3.97  595414640   
2           NaN    zinger   3.97  595412617   
3           NaN       NaN   4.92  420652863   
4           NaN       NaN   3.92  484071203   

                           user_session  cart  purchase  remove_from_cart  \
0  4adb70bb-edbd-4981-b60f-a05bfd32683a     0         0                 0   
1  c8c5205d-be43-4f1d-aa56-4828b8151c8a     0         0                 0   
2  46a5010f-bd69-4fbe-a00d-bb17aa7b46f3     0         0                 0   
3  546f6af3-a517-4752-a98b-80c4c5860711     0         0                 0   
4  cff70ddf-529e-4b0c-a4fc-f43a749c0acb     0         0                 0   

   view  
0     1  
1     1  
2     1  
3     1  
4     1

In [10]:
cos.price.max()

327.78

**Converting times into different format and add them as new column**

Creating time column regard to date, specific timepoint, and week

In [11]:
%time
from datetime import datetime
import matplotlib.dates as dates
%matplotlib inline


cos['date'] = cos['event_time'].apply(lambda r: datetime.strptime(str(r)[:10],'%Y-%m-%d'))
cos['timepoint'] = cos['event_time'].apply(lambda r: datetime.strptime(str(r)[11:19], '%H:%M:%S').time())
cos['week'] = cos['date'].apply(lambda d: (d.day-1) // 7 + 1)

cos.head()

CPU times: user 2 µs, sys: 1e+03 ns, total: 3 µs
Wall time: 4.29 µs


event_time event_type  product_id          category_id  \
0 2020-01-01 00:00:00+00:00       view     5809910  1602943681873052386   
1 2020-01-01 00:00:09+00:00       view     5812943  1487580012121948301   
2 2020-01-01 00:00:19+00:00       view     5798924  1783999068867920626   
3 2020-01-01 00:00:24+00:00       view     5793052  1487580005754995573   
4 2020-01-01 00:00:25+00:00       view     5899926  2115334439910245200   

  category_code     brand  price    user_id  \
0           NaN   grattol   5.24  595414620   
1           NaN  kinetics   3.97  595414640   
2           NaN    zinger   3.97  595412617   
3           NaN       NaN   4.92  420652863   
4           NaN       NaN   3.92  484071203   

                           user_session  cart  purchase  remove_from_cart  \
0  4adb70bb-edbd-4981-b60f-a05bfd32683a     0         0                 0   
1  c8c5205d-be43-4f1d-aa56-4828b8151c8a     0         0                 0   
2  46a5010f-bd69-4fbe-a00d-bb17aa7b46f3     0         0                 0   
3  546f6af3-a517-4752-a98b-80c4c5860711     0         0                 0   
4  cff70ddf-529e-4b0c-a4fc-f43a749c0acb     0         0                 0   

   view       date timepoint  week  
0     1 2020-01-01  00:00:00     1  
1     1 2020-01-01  00:00:09     1  
2     1 2020-01-01  00:00:19     1  
3     1 2020-01-01  00:00:24     1  
4     1 2020-01-01  00:00:25     1

In [12]:
cos['category_code'].dropna()

154                      accessories.bag
359                  stationery.cartrige
380                  stationery.cartrige
487        appliances.environment.vacuum
506        appliances.environment.vacuum
                       ...              
4264525         accessories.cosmetic_bag
4264534         accessories.cosmetic_bag
4264594    furniture.living_room.cabinet
4264721                    apparel.glove
4264738    appliances.environment.vacuum
Name: category_code, Length: 74719, dtype: object

In [13]:
pd.isnull(cos['category_code'][155])

True

From the cell above, we can see that there's non-empty category_code for some event. 

And those non-empty category code **follows the pattern with [class].[item].**

Since the analysis will be more focused on the class perspective, I'll remove the specific item and only leave the category.

In [15]:
import re

# a function to get rid of items at the end of category code
#  matches one period in first 30 chars,
#  drops everything after that match (plus the match itself)
def remove_itemnames(text):
    '''
    Remove specific item string from a text data
    If there is a period in the first 30 characters, 
      drops everything after the period(including the period it self)
    If no hyphen/colon, do nothing
    Return processed string
    '''
    if pd.isnull(text) is False:
        pos = text.find('.')
        if pos != -1:
            return text[:pos]
        else:
            return text
    else:
        return text

In [16]:
a= '1234'
a[-1]

'4'

In [17]:
# testing the method
new = cos['category_code'].apply(remove_itemnames)
new.dropna()

154        accessories
359         stationery
380         stationery
487         appliances
506         appliances
              ...     
4264525    accessories
4264534    accessories
4264594      furniture
4264721        apparel
4264738     appliances
Name: category_code, Length: 74719, dtype: object

In [18]:
# applying the method of removing the specific item names to the original dataframe
cos['category_code'] = cos['category_code'].apply(remove_itemnames)


In [19]:
# displaying the dataset after cleaning
cos.head()

event_time event_type  product_id          category_id  \
0 2020-01-01 00:00:00+00:00       view     5809910  1602943681873052386   
1 2020-01-01 00:00:09+00:00       view     5812943  1487580012121948301   
2 2020-01-01 00:00:19+00:00       view     5798924  1783999068867920626   
3 2020-01-01 00:00:24+00:00       view     5793052  1487580005754995573   
4 2020-01-01 00:00:25+00:00       view     5899926  2115334439910245200   

  category_code     brand  price    user_id  \
0           NaN   grattol   5.24  595414620   
1           NaN  kinetics   3.97  595414640   
2           NaN    zinger   3.97  595412617   
3           NaN       NaN   4.92  420652863   
4           NaN       NaN   3.92  484071203   

                           user_session  cart  purchase  remove_from_cart  \
0  4adb70bb-edbd-4981-b60f-a05bfd32683a     0         0                 0   
1  c8c5205d-be43-4f1d-aa56-4828b8151c8a     0         0                 0   
2  46a5010f-bd69-4fbe-a00d-bb17aa7b46f3     0         0                 0   
3  546f6af3-a517-4752-a98b-80c4c5860711     0         0                 0   
4  cff70ddf-529e-4b0c-a4fc-f43a749c0acb     0         0                 0   

   view       date timepoint  week  
0     1 2020-01-01  00:00:00     1  
1     1 2020-01-01  00:00:09     1  
2     1 2020-01-01  00:00:19     1  
3     1 2020-01-01  00:00:24     1  
4     1 2020-01-01  00:00:25     1

**I'll take a look into the column 'price'**

In [20]:
cos['price'].describe()

count    4.264752e+06
mean     8.713981e+00
std      1.963959e+01
min     -7.937000e+01
25%      2.350000e+00
50%      4.060000e+00
75%      7.140000e+00
max      3.277800e+02
Name: price, dtype: float64

In [21]:
# I'll take a look into the min price, which is abnormally less than zero
cos.price.min()


-79.37

In [22]:
abnormalprice = cos.loc[cos.price < 0]
abnormalprice.head()

event_time event_type  product_id          category_id  \
209963 2020-01-03 10:21:27+00:00   purchase     5716859  1487580014042939619   
298228 2020-01-03 22:49:36+00:00   purchase     5716861  1487580014042939619   
450012 2020-01-05 10:49:53+00:00       view     5716857  1487580014042939619   
455893 2020-01-05 11:40:45+00:00   purchase     5716857  1487580014042939619   
459678 2020-01-05 12:11:40+00:00   purchase     5670257  1487580014042939619   

       category_code brand  price    user_id  \
209963           NaN   NaN -47.62  577738817   
298228           NaN   NaN -79.37  560746323   
450012           NaN   NaN -23.81  439078759   
455893           NaN   NaN -23.81  583559162   
459678           NaN   NaN -15.87  427985196   

                                user_session  cart  purchase  \
209963  140d150d-abd0-4202-b6a5-ce0525b2ced4     0         1   
298228  024093b7-5d73-44a7-8eff-c9fe0b21fcf3     0         1   
450012  4a07f317-1335-4706-ac80-5a25a9c22581     0         0   
455893  7789d1b2-6658-4e5a-beb7-627c0e1ed6e0     0         1   
459678  39a99725-2a02-4191-835e-7356c8055fa7     0         1   

        remove_from_cart  view       date timepoint  week  
209963                 0     0 2020-01-03  10:21:27     1  
298228                 0     0 2020-01-03  22:49:36     1  
450012                 0     1 2020-01-05  10:49:53     1  
455893                 0     0 2020-01-05  11:40:45     1  
459678                 0     0 2020-01-05  12:11:40     1

Usually the price would be positive. In this particular scene, I first thought of having discount might lead to specific price. However, the data given above shows very little possibility of discount, which usually happens on specific holiday, for particular brand, etc. 

Thus, I'll assume that these negative price is made by mistake while manually put in the data for analysis. I'll convert them into positive number.



In [23]:
# converting negative price into positive
cos['price']=abs(cos['price'])


In [24]:
# Examining if the converting process is successful
cos.loc[cos.price<0]
# Yes, it is!!:)

Empty DataFrame
Columns: [event_time, event_type, product_id, category_id, category_code, brand, price, user_id, user_session, cart, purchase, remove_from_cart, view, date, timepoint, week]
Index: []

In [25]:
# Rechecking the price column
cos.price.describe()

count    4.264752e+06
mean     8.714595e+00
std      1.963932e+01
min      0.000000e+00
25%      2.350000e+00
50%      4.060000e+00
75%      7.140000e+00
max      3.277800e+02
Name: price, dtype: float64

In [26]:
len(cos.price)

4264752

**Turning time into categorical variables**

In [45]:
def get_sj(x):
    time_6=datetime.strptime('6:00:00', '%H:%M:%S').time()
    time_12=datetime.strptime('12:00:00', '%H:%M:%S').time()
    time_18=datetime.strptime('18:00:00', '%H:%M:%S').time()
    xtime = x#.timetz()
    if xtime<time_6:
        return 'wee_hours'
    elif xtime<time_12:
        return 'morning'
    elif xtime<time_18:
        return 'afternoon'
    else:
        return 'evening'


In [46]:
cos['timeblocks'] = cos['timepoint'].map(get_sj)

In [47]:
cos['timeblocks']

0          wee_hours
1          wee_hours
2          wee_hours
3          wee_hours
4          wee_hours
             ...    
4264747      evening
4264748      evening
4264749      evening
4264750      evening
4264751      evening
Name: timeblocks, Length: 4264752, dtype: object

**using .get_dummies function to add more blocks of timepoint**

I'll turn these into categorical variables in order to make it easier for fitting the logistics regression model.

In [49]:
cos['is_weehours']=pd.get_dummies(cos['timeblocks'])['wee_hours']
cos['is_morning']=pd.get_dummies(cos['timeblocks'])['morning']
cos['is_afternoon']=pd.get_dummies(cos['timeblocks'])['afternoon']
cos['is_evening']=pd.get_dummies(cos['timeblocks'])['evening']

cos.head()



event_time event_type  product_id          category_id  \
0 2020-01-01 00:00:00+00:00       view     5809910  1602943681873052386   
1 2020-01-01 00:00:09+00:00       view     5812943  1487580012121948301   
2 2020-01-01 00:00:19+00:00       view     5798924  1783999068867920626   
3 2020-01-01 00:00:24+00:00       view     5793052  1487580005754995573   
4 2020-01-01 00:00:25+00:00       view     5899926  2115334439910245200   

  category_code     brand  price    user_id  \
0           NaN   grattol   5.24  595414620   
1           NaN  kinetics   3.97  595414640   
2           NaN    zinger   3.97  595412617   
3           NaN       NaN   4.92  420652863   
4           NaN       NaN   3.92  484071203   

                           user_session  cart  ...  remove_from_cart  view  \
0  4adb70bb-edbd-4981-b60f-a05bfd32683a     0  ...                 0     1   
1  c8c5205d-be43-4f1d-aa56-4828b8151c8a     0  ...                 0     1   
2  46a5010f-bd69-4fbe-a00d-bb17aa7b46f3     0  ...                 0     1   
3  546f6af3-a517-4752-a98b-80c4c5860711     0  ...                 0     1   
4  cff70ddf-529e-4b0c-a4fc-f43a749c0acb     0  ...                 0     1   

        date timepoint week  timeblocks is_weehours  is_morning  is_afternoon  \
0 2020-01-01  00:00:00    1   wee_hours           1           0             0   
1 2020-01-01  00:00:09    1   wee_hours           1           0             0   
2 2020-01-01  00:00:19    1   wee_hours           1           0             0   
3 2020-01-01  00:00:24    1   wee_hours           1           0             0   
4 2020-01-01  00:00:25    1   wee_hours           1           0             0   

   is_evening  
0           0  
1           0  
2           0  
3           0  
4           0  

[5 rows x 21 columns]



**Written the modified dataset into a .csv file**

In [50]:
cos.to_csv("finaldataset.csv")

In [53]:
cos['category_code'].dropna()

154        accessories
359         stationery
380         stationery
487         appliances
506         appliances
              ...     
4264525    accessories
4264534    accessories
4264594      furniture
4264721        apparel
4264738     appliances
Name: category_code, Length: 74719, dtype: object